File: packages/daemon/src/logging/logger.ts
/**
 * Logger class for structured logging in the daemon
 */

import type { LogLevel, LogEntry } from '@ghidra-mcp/shared';
import { LOG_LEVEL_PRIORITY } from '@ghidra-mcp/shared';
import type { LogStore } from './store.js';

export interface LoggerOptions {
  component: string;
  store: LogStore;
  sessionId?: string;
  workerId?: string;
  minLevel?: LogLevel;
}

export class Logger {
  private readonly component: string;
  readonly store: LogStore;
  private readonly sessionId?: string;
  private readonly workerId?: string;
  private minLevel: LogLevel;

  constructor(options: LoggerOptions) {
    this.component = options.component;
    this.store = options.store;
    this.sessionId = options.sessionId;
    this.workerId = options.workerId;
    this.minLevel = options.minLevel ?? 'INFO';
  }

  /**
   * Set the minimum log level
   */
  setLevel(level: LogLevel): void {
    this.minLevel = level;
  }

  /**
   * Log an error message
   */
  error(message: string, metadata?: Record<string, unknown>): void {
    this.log('ERROR', message, metadata);
  }

  /**
   * Log a warning message
   */
  warn(message: string, metadata?: Record<string, unknown>): void {
    this.log('WARN', message, metadata);
  }

  /**
   * Log an info message
   */
  info(message: string, metadata?: Record<string, unknown>): void {
    this.log('INFO', message, metadata);
  }

  /**
   * Log a debug message
   */
  debug(message: string, metadata?: Record<string, unknown>): void {
    this.log('DEBUG', message, metadata);
  }

  /**
   * Create a child logger with additional context
   */
  child(options: { sessionId?: string; workerId?: string; component?: string }): Logger {
    return new Logger({
      component: options.component ?? this.component,
      store: this.store,
      sessionId: options.sessionId ?? this.sessionId,
      workerId: options.workerId ?? this.workerId,
      minLevel: this.minLevel,
    });
  }

  /**
   * Internal log method
   */
  private log(level: LogLevel, message: string, metadata?: Record<string, unknown>): void {
    // Check if we should log at this level
    if (LOG_LEVEL_PRIORITY[level] > LOG_LEVEL_PRIORITY[this.minLevel]) {
      return;
    }

    const entry: LogEntry = {
      timestamp: Date.now(),
      level,
      source: 'daemon',
      component: this.component,
      sessionId: this.sessionId,
      workerId: this.workerId,
      message,
      metadata,
    };

    // Store the log entry
    this.store.append(entry);

    // Also output to console for backwards compatibility and debugging
    this.consoleOutput(entry);
  }

  /**
   * Output to console in a formatted way
   */
  private consoleOutput(entry: LogEntry): void {
    const timestamp = new Date(entry.timestamp).toISOString();
    const prefix = `[${timestamp}][${entry.component}][${entry.level}]`;

    let contextParts: string[] = [];
    if (entry.sessionId) {
      contextParts.push(`session=${entry.sessionId.slice(0, 8)}`);
    }
    if (entry.workerId) {
      contextParts.push(`worker=${entry.workerId.slice(0, 8)}`);
    }

    const context = contextParts.length > 0 ? ` (${contextParts.join(', ')})` : '';
    const metadataStr = entry.metadata ? ` ${JSON.stringify(entry.metadata)}` : '';

    const output = `${prefix}${context} ${entry.message}${metadataStr}`;

    switch (entry.level) {
      case 'ERROR':
        console.error(output);
        break;
      case 'WARN':
        console.warn(output);
        break;
      case 'DEBUG':
        console.debug(output);
        break;
      default:
        console.log(output);
    }
  }
}

/**
 * Create a root logger for the daemon
 */
export function createLogger(store: LogStore, component: string): Logger {
  return new Logger({
    component,
    store,
    minLevel: process.env.LOG_LEVEL as LogLevel ?? 'INFO',
  });
}

File: ghidra-worker/src/main/java/com/ghidramcp/logging/Logger.java
package com.ghidramcp.logging;

import com.google.gson.Gson;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * Logger for the Ghidra worker that streams logs to the daemon via WebSocket.
 * Logs are also printed to stdout for backwards compatibility and local debugging.
 */
public class Logger {

    public enum Level {
        ERROR(0),
        WARN(1),
        INFO(2),
        DEBUG(3);

        private final int priority;

        Level(int priority) {
            this.priority = priority;
        }

        public int getPriority() {
            return priority;
        }
    }

    private static final Gson gson = new Gson();
    private static final Object pendingLock = new Object();
    private static final List<LogEntry> pending = new ArrayList<>();
    private static WebSocket webSocket;
    private static ScheduledExecutorService flushScheduler;
    private static boolean initialized = false;
    private static String globalWorkerId;
    private static String globalSessionId;

    private final String component;
    private final String sessionId;
    private final String workerId;
    private Level minLevel = Level.INFO;

    /**
     * Create a new Logger instance.
     */
    public Logger(String component, String workerId, String sessionId) {
        this.component = component;
        this.workerId = workerId;
        this.sessionId = sessionId;
    }

    /**
     * Initialize the WebSocket connection to the daemon.
     * Should be called once at worker startup.
     */
    public static void initWebSocket(String daemonUrl, String workerId, String sessionId) {
        if (initialized) {
            return;
        }

        globalWorkerId = workerId;
        globalSessionId = sessionId;

        try {
            // Convert http:// to ws:// or https:// to wss://
            String secret = System.getenv("GHIDRA_MCP_WORKER_SECRET");
            String wsUrl = daemonUrl.replace("http://", "ws://").replace("https://", "wss://")
                    + "/internal/ws/logs?workerId=" + workerId
                    + (secret != null ? "&secret=" + java.net.URLEncoder.encode(secret, java.nio.charset.StandardCharsets.UTF_8) : "");

            HttpClient client = HttpClient.newBuilder()
                    .connectTimeout(Duration.ofSeconds(5))
                    .build();

            webSocket = client.newWebSocketBuilder()
                    .buildAsync(URI.create(wsUrl), new LogWebSocketListener())
                    .join();

            // Start flush thread (every 100ms)
            flushScheduler = Executors.newSingleThreadScheduledExecutor(r -> {
                Thread t = new Thread(r, "log-flush-thread");
                t.setDaemon(true);
                return t;
            });
            flushScheduler.scheduleAtFixedRate(Logger::flush, 100, 100, TimeUnit.MILLISECONDS);

            initialized = true;
            System.out.println("[Logger] WebSocket logging initialized");
        } catch (Exception e) {
            System.err.println("[Logger] Failed to initialize WebSocket logging: " + e.getMessage());
            // Continue without WebSocket - logs will still go to stdout
        }
    }

    /**
     * Shutdown the WebSocket connection and flush remaining logs.
     */
    public static void shutdown() {
        if (flushScheduler != null) {
            flushScheduler.shutdown();
            try {
                flushScheduler.awaitTermination(1, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        // Final flush
        flush();

        if (webSocket != null) {
            webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "shutdown");
        }

        initialized = false;
    }

    /**
     * Create a root logger instance.
     */
    public static Logger create(String component, String workerId, String sessionId) {
        return new Logger(component, workerId, sessionId);
    }

    /**
     * Set the minimum log level.
     */
    public void setLevel(Level level) {
        this.minLevel = level;
    }

    /**
     * Create a child logger with a different component name.
     */
    public Logger child(String childComponent) {
        Logger child = new Logger(childComponent, this.workerId, this.sessionId);
        child.minLevel = this.minLevel;
        return child;
    }

    /**
     * Log an error message.
     */
    public void error(String message) {
        log(Level.ERROR, message, null);
    }

    /**
     * Log an error message with metadata.
     */
    public void error(String message, Map<String, Object> metadata) {
        log(Level.ERROR, message, metadata);
    }

    /**
     * Log a warning message.
     */
    public void warn(String message) {
        log(Level.WARN, message, null);
    }

    /**
     * Log a warning message with metadata.
     */
    public void warn(String message, Map<String, Object> metadata) {
        log(Level.WARN, message, metadata);
    }

    /**
     * Log an info message.
     */
    public void info(String message) {
        log(Level.INFO, message, null);
    }

    /**
     * Log an info message with metadata.
     */
    public void info(String message, Map<String, Object> metadata) {
        log(Level.INFO, message, metadata);
    }

    /**
     * Log a debug message.
     */
    public void debug(String message) {
        log(Level.DEBUG, message, null);
    }

    /**
     * Log a debug message with metadata.
     */
    public void debug(String message, Map<String, Object> metadata) {
        log(Level.DEBUG, message, metadata);
    }

    /**
     * Internal log method.
     */
    private void log(Level level, String message, Map<String, Object> metadata) {
        if (level.getPriority() > minLevel.getPriority()) {
            return;
        }

        LogEntry entry = new LogEntry();
        entry.timestamp = System.currentTimeMillis();
        entry.level = level.name();
        entry.source = "worker";
        entry.component = component;
        entry.sessionId = sessionId;
        entry.workerId = workerId;
        entry.message = message;
        entry.metadata = metadata;

        // Add to pending queue for WebSocket transmission
        synchronized (pendingLock) {
            pending.add(entry);
        }

        // Also print to stdout (backwards compat + local debugging)
        printToConsole(entry);
    }

    /**
     * Print a log entry to the console.
     */
    private void printToConsole(LogEntry entry) {
        String prefix = String.format("[%s][%s]", entry.component, entry.level);
        String metaStr = entry.metadata != null ? " " + gson.toJson(entry.metadata) : "";
        String output = prefix + " " + entry.message + metaStr;

        if ("ERROR".equals(entry.level)) {
            System.err.println(output);
        } else {
            System.out.println(output);
        }
    }

    /**
     * Flush pending logs to the WebSocket.
     */
    private static void flush() {
        List<LogEntry> toSend;
        synchronized (pendingLock) {
            if (pending.isEmpty()) {
                return;
            }
            toSend = new ArrayList<>(pending);
            pending.clear();
        }

        if (webSocket != null) {
            try {
                String json = gson.toJson(toSend);
                webSocket.sendText(json, true);
            } catch (Exception e) {
                // Log to stderr but don't throw - we don't want logging to crash the worker
                System.err.println("[Logger] Failed to send logs: " + e.getMessage());
            }
        }
    }

    /**
     * WebSocket listener for the log connection.
     */
    private static class LogWebSocketListener implements WebSocket.Listener {
        @Override
        public void onOpen(WebSocket ws) {
            System.out.println("[Logger] WebSocket connection opened");
        }

        @Override
        public CompletionStage<?> onClose(WebSocket ws, int statusCode, String reason) {
            System.out.println("[Logger] WebSocket connection closed: " + statusCode + " - " + reason);
            return null;
        }

        @Override
        public void onError(WebSocket ws, Throwable error) {
            System.err.println("[Logger] WebSocket error: " + error.getMessage());
        }

        @Override
        public CompletionStage<?> onText(WebSocket ws, CharSequence data, boolean last) {
            // We don't expect to receive messages, but request more if we do
            ws.request(1);
            return null;
        }
    }
}
